### PR TITLE
Dynamically Loading Libraries

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Refactors the modules and packages systems so that they can be loaded dynamically.

Related to #1781, but does not implement its full spec (edition resolution is left for a separate PR).

It also implements an interface allowing to use the standard SLF4J logging utils from within the `runtime`; this interface is connected to the `TruffleLogger` which then passes the logs to our logging service.

### Important Notes

- I removed the concept of 'shadowed packages' as it does not seem to be well-defined in the context of the new edition-based resolution system, but I'm willing to discuss if this should be handled differently. Main point is - we do not know what packages are available (well, we could look through what is defined in the edition + what is available on all paths in the local search path) so it is unclear if something is shadowing something else. 
  - The only reasonable situation seems to report that the current project is shadowing some library that is defined within the edition - this could indeed be checked.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Comment by radeusgd** [comment]
> * I removed the concept of 'shadowed packages' as it does not seem to be well-defined in the context of the new edition-based resolution system, but I'm willing to discuss if this should be handled differently. Main point is - we do not know what packages are available (well, we could look through what is defined in the edition + what is available on all paths in the local search path) so it is unclear if something is shadowing something else.
>   
>   * The only reasonable situation seems to report that the current project is shadowing some library that is defined within the edition - this could indeed be checked.

cc: @kustosz @iamrecursion any thoughts about this?

**Review by kustosz — CHANGES_REQUESTED** [review]
Code looks good, but there's an architectural problem with this solution. You have the context (which is "deep compiler, here be dragons") talking directly to the language server (which is "what compiler, me just want code go brrr"). This skips a layer of our runtime server, responsible for translating between the compiler and the LS. This will probably result in very client-specific logic being added to the context over time. All of this logic should be mediated by our pre-existing handlers system.

**Review by iamrecursion — COMMENTED** [review]
Just some notes on documentation. I'd like to echo what Marcin has said in his review.

**Review by kustosz — APPROVED** [review]
LGTM

File: engine/runtime/src/main/java/org/enso/interpreter/Language.java [hunk 1]
```diff
@@ -60,11 +63,18 @@
    */
   @Override
   protected Context createContext(Env env) {
-    Context context = new Context(this, getLanguageHome(), env);
+    var notificationHandler = new Forwarder();
+    boolean isInteractiveMode = env.getOptions().get(RuntimeOptions.INTERACTIVE_MODE_KEY);
+    boolean isTextMode = !isInteractiveMode;
+    if (!isTextMode) {
```
Comment (4e6) [inline engine/runtime/src/main/java/org/enso/interpreter/Language.java]: are you just inverted interactive mode twice, or I'm missing something?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java [hunk 2]
```diff
@@ -145,15 +118,12 @@ private static Module getModule(
...
-      }
-      Module module = scope.modules.get(moduleName);
-      if (module == null) {
+      var module = scope.getModule(moduleName);
+      if (module.isEmpty()) {
         throw UnknownIdentifierException.create(moduleName);
       }
 
-      return module;
+      return module.get();
```
Comment (4e6) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java:126]: I'd throw a custom error here.

**Review by 4e6 — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/Language.java (hunk 1 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/Language.java]: Oh that is a mistake, it was supposed to be `if(isTextMode)`! Thanks!

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java (hunk 2 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java:126]: There is a check if the option is filled above.

**Pull request merged**